site/src/components/MultiSelectCombobox/MultiSelectCombobox.tsx

```diff
@@ -0,0 +1,640 @@
+function isOptionsExist(groupOption: GroupOption, targetOption: Option[]) {
+	return Object.values(groupOption).some((value) =>
+		value.some((option) => targetOption.some((o) => o.value === option.value)),
+	);
+}
+
+/**
+ * The `CommandEmpty` of shadcn/ui will cause the cmdk-empty to not render correctly.
+ * Here a new CommandEmpty is created using the `Empty` implementation from `cmdk`.
+ *
+ * @reference: https://github.com/hsuanyi-chou/shadcn-ui-expansions/issues/34#issuecomment-1949561607
+ **/
+const CommandEmpty = forwardRef<
+	HTMLDivElement,
+	React.ComponentProps<typeof CommandPrimitive.Empty>
+>(({ className, ...props }, forwardedRef) => {
+	const render = useCommandState((state) => state.filtered.count === 0);
+
+	if (!render) return null;
+
+	return (
+		<div
+			ref={forwardedRef}
+			className={cn("py-6 text-center text-sm", className)}
+			cmdk-empty=""
+			role="presentation"
+			{...props}
+		/>
+	);
+});
+
+export const MultiSelectCombobox = React.forwardRef<
+	MultiSelectComboboxRef,
+	MultiSelectComboboxProps
+>(
+	(
+		{
+			value,
+			onChange,
+			placeholder,
+			defaultOptions: arrayDefaultOptions = [],
+			options: arrayOptions,
+			delay,
+			onSearch,
+			onSearchSync,
+			loadingIndicator,
+			emptyIndicator,
+			maxSelected = Number.MAX_SAFE_INTEGER,
+			onMaxSelected,
+			hidePlaceholderWhenSelected,
+			disabled,
+			groupBy,
+			className,
+			badgeClassName,
+			selectFirstItem = true,
+			creatable = false,
+			triggerSearchOnFocus = false,
+			commandProps,
+			inputProps,
+			hideClearAllButton = false,
+		}: MultiSelectComboboxProps,
+		ref,
+	) => {
+		const inputRef = React.useRef<HTMLInputElement>(null);
+		const [open, setOpen] = React.useState(false);
+		const [onScrollbar, setOnScrollbar] = React.useState(false);
+		const [isLoading, setIsLoading] = React.useState(false);
+		const dropdownRef = React.useRef<HTMLDivElement>(null); // Added this
+
+		const [selected, setSelected] = React.useState<Option[]>(value || []);
+		const [options, setOptions] = React.useState<GroupOption>(
+			transitionToGroupOption(arrayDefaultOptions, groupBy),
+		);
+		const [inputValue, setInputValue] = React.useState("");
+		const debouncedSearchTerm = useDebouncedValue(inputValue, delay || 500);
+
+		const [previousValue, setPreviousValue] = React.useState<Option[]>(
+			value || [],
+		);
+		if (value && value !== previousValue) {
+			setPreviousValue(value);
+			setSelected(value);
+		}
+
+		React.useImperativeHandle(
+			ref,
+			() => ({
+				selectedValue: [...selected],
+				input: inputRef.current as HTMLInputElement,
+				focus: () => inputRef?.current?.focus(),
+				reset: () => setSelected([]),
+			}),
+			[selected],
+		);
+
+		const handleUnselect = React.useCallback(
+			(option: Option) => {
+				const newOptions = selected.filter((s) => s.value !== option.value);
+				setSelected(newOptions);
+				onChange?.(newOptions);
+			},
+			[onChange, selected],
+		);
+
+		const handleKeyDown = React.useCallback(
```

Comment:
These `useCallback` calls only add overhead. They're not used anywhere where effects or memoization matters, and the new functions are still getting recreated every render – they're just getting thrown away sometimes